code/modules/research/xenoarchaeology/tools/bunsen_burner.dm

```diff
@@ -5,56 +5,167 @@
 	icon = 'icons/obj/device.dmi'
 	icon_state = "bunsen0"
 	var/heating = 0		//whether the bunsen is turned on
-	var/heated = 0		//whether the bunsen has been on long enough to let stuff react
 	var/obj/item/weapon/reagent_containers/held_container
-	var/heat_time = 50
+	var/list/possible_fuels = list(
+		PLASMA = list(
+				"max_temperature" = TEMPERATURE_PLASMA,
+				"thermal_energy_transfer" = 6000,
+				"consumption_rate" = 0.1,
+				"o2_cons" = 0.01,
+				"co2_cons" = 0,
+				"unsafety" = 0),
+		GLYCEROL = list(
+				"max_temperature" = 1833.15,
+				"thermal_energy_transfer" = 4000,
+				"consumption_rate" = 0.25,
+				"o2_cons" = 0.05,
+				"co2_cons" = -0.025,
+				"unsafety" = 5),
+		FUEL = list(
+				"max_temperature" = TEMPERATURE_WELDER,
+				"thermal_energy_transfer" = 3600,
+				"consumption_rate" = 0.5,
+				"o2_cons" = 0.2,
+				"co2_cons" = -0.2,
+				"unsafety" = 25),
+		ETHANOL = list(
+				"max_temperature" = 1833.15,
+				"thermal_energy_transfer" = 2600,
+				"consumption_rate" = 0.5,
+				"o2_cons" = 0.08,
+				"co2_cons" = -0.04,
+				"unsafety" = 10))
 	pixel_y = 15 * PIXEL_MULTIPLIER
 	ghost_read = 0
 
-/obj/machinery/bunsen_burner/attackby(obj/item/weapon/W as obj, mob/user as mob)
+/obj/machinery/bunsen_burner/New()
+	..()
+	processing_objects.Remove(src)
+	create_reagents(50)
+
+/obj/machinery/bunsen_burner/Destroy()
+	if(held_container)
+		held_container.forceMove(get_turf(src))
+		held_container = null
+	..()
+
+/obj/machinery/bunsen_burner/attackby(obj/item/weapon/W, mob/user)
 	if(istype(W, /obj/item/weapon/reagent_containers))
+		var/obj/item/weapon/reagent_containers/R = W
 		if(held_container)
+			for(var/possible_fuel in possible_fuels)
+				if(R.reagents.has_reagent(possible_fuel) && (reagents.has_reagent(possible_fuel) || !reagents.reagent_list.len))
+					var/reagent_transfer = R.reagents.trans_id_to(src, possible_fuel, 10)
+					if(reagent_transfer)
+						to_chat(user, "<span class='notice'>You transfer [reagent_transfer]u of [possible_fuel] from \the [R] to \the [src]</span>")
+						return
 			to_chat(user, "<span class='warning'>You must remove the [held_container] first.</span>")
+			return
 		else
 			if(user.drop_item(W, src))
 				held_container = W
 				to_chat(user, "<span class='notice'>You put the [held_container] onto the [src].</span>")
 				var/image/I = image("icon"=W, "layer"=FLOAT_LAYER)
 				underlays += I
-				if(heating)
-					spawn(heat_time)
-						try_heating()
-
 				return 1 // avoid afterattack() being called
+	if(iswrench(W))
+		user.visible_message("<span class = 'warning'>[user] starts to deconstruct \the [src]!</span>","<span class = 'notice'>You start to deconstruct \the [src].</span>")
+		if(do_after(user, src, 5 SECONDS))
+			playsound(get_turf(src), 'sound/items/Ratchet.ogg', 50, 1)
+			drop_stack(/obj/item/stack/sheet/metal, loc, rand(3,4), user)
+			qdel(src)
+			return
```

Comment:
The `return` here isn't necessary, all remaining code in this proc is inside an `else` statement, and so could not be executed if this code is.